[DDCI-246] adhoc audit report

awset opened the pull request:
https://it-partners.atlassian.net/browse/DDCI-246


![image](https://user-images.githubusercontent.com/1538818/98910102-4d747f80-24f5-11eb-821c-bebe8b819911.png)

salsa-nathan commented on ckanext/qdes/logic/action/get.py:
```diff
@@ -0,0 +1,341 @@
+    """
+    # Get list of invalid uris.
+    uris = Session.query(InvalidUri).all()
+    invalid_uris = [uri.as_dict() for uri in uris]
```
Hi @awang - converting each of the `InvalidUri` objects may not be necessary here as we can just iterate through the objects in the `for invalid_uri in invalid_uris:` block below and use the object's properties to populate the `entities` dict... I will add an example below.

awset replied:
ok noted, will implement for others.

salsa-nathan commented on ckanext/qdes/logic/action/get.py:
```diff
@@ -0,0 +1,341 @@
+        if invalid_uri.get('entity_id') in entities:
+            entities[invalid_uri.get('entity_id')].get('fields').append(invalid_uri.get('field'))
+        else:
+            entities[invalid_uri.get('entity_id')] = {}
```
Hi @awset - we don't need to initialise an empty dict here then `.update(...)` it - we can just add the dict straight to `entities` like:
```
entities[invalid_uri.get('entity_id')] = {
                'type': invalid_uri.get('entity_type'),
                'fields': [invalid_uri.get('field')],
            }
```

awset replied:
ok noted, will implement for others.

salsa-nathan commented on ckanext/qdes/logic/action/get.py:
```diff
@@ -0,0 +1,341 @@
+            entities[invalid_uri.get('entity_id')].update({
+                'type': invalid_uri.get('entity_type'),
+                'fields': [invalid_uri.get('field')],
+            })
```
Hi @awset - this block (from `# Get list of invalid uris.` down to `# Build rows.`) can be simplified for example:
```
    # Get list of invalid uris.
    invalid_uris = Session.query(InvalidUri).all()

    # Build package list.
    entities = {}
    for uri in invalid_uris:
        if uri.entity_id in entities:
            entities[uri.entity_id]['fields'].append(uri.field)
        else:
            entities[uri.entity_id] = {
                'type': uri.entity_type,
                'fields': [uri.field],
            }
```
...makes the code easier to read & maintain. It could possibly be extracted to a helper function too, to reduce the function size for maintainability down the track

awset replied:
I have moved it to helper function

salsa-nathan commented on ckanext/qdes/logic/action/get.py:
```diff
@@ -0,0 +1,341 @@
+            else:
+                entity_dict = get_action('package_show')({}, {'id': entity_id})
+
+                if config and config.get('org_id') and entity_dict.get('organization').get('id') != config.get(
```
Hi @awset - `config.get('org_id')` is used multiple times in this function... if we default set `config={}` in the function declaration, then set `org_id = config.get('org_id', None)` at the start of this function, we can just refer to `org_id` var within the function and reduce the `if config and config.get('org_id')` checks to `if org_id`

awset replied:
I will implement it for other functions

salsa-nathan commented on ckanext/qdes/logic/action/get.py:
```diff
@@ -0,0 +1,341 @@
+                    continue
+
+                # Cache the result.
+                packages[entity_id] = {}
```
Hi @awset as per comment above - we don't need to initialise an empty dict here then .update(...) it - we can just add the dict straight to `packages`, e.g.
```
packages[entity_id] = entity_dict
```

awset replied:
noted

salsa-nathan commented on ckanext/qdes/logic/action/get.py:
```diff
@@ -0,0 +1,341 @@
+        # Load entity and cache them.
+        entity_dict = {}
+        parent_entity_dict = {}
+        invalid_uri = entities.get(entity_id)
```
Hi @awset - we can avoid having to get the entity value from the `entities` dict here, by iterating through the `entities` items, e.g. change:
```
for entity_id in entities:
```
to:
```
for entity_id, invalid_uri in entities.items():
```
(e.g. `for key, value in entities.items():`)

Then we can remove this line (`invalid_uri = entities.get(entity_id)`) as we will already have `invalid_url` (value) set

awset replied:
noted, I still not really fluent python list/object/dict manipulation

salsa-nathan replied:
@awset that's all good man - it all takes time, and there's always lots of different ways to do things :)

salsa-nathan commented:
Hi @awset 

Apologies for all the comments.

I have submitted a PR with proposed changes for the `qdes_datasets_with_invalid_urls`

https://github.com/salsadigitalauorg/ckanext-qdes/pull/5

I think we need to implement similar changes to the other functions in `get.py` too, such as:

- each function in the `get.py` is accessible via the API, e.g. http://qdes-ckan-29.docker.amazee.io/api/action/qdes_datasets_with_invalid_urls -- so we need to lock these down to `sysadmin` users only (see example in PR above)
- calling `_qdes_extract_point_of_contact` to get a particular property (Name, Email, etc) multiple times will create a CSV lookup each time - this could take up a lot of time if that vocab is large (which it probably will be) - instead we should look it up once and store it in a local var (i.e. see the `get_point_of_contact` function in `report_helpers.py` in PR above) then get the value from the local var when we need to use it say in the CSV row
- functions like `_qdes_get_organization_dict_by_id`, `_get_recommended_dataset_fields`, `_get_uri_validated_fields` should be moved into a helper file if they don't need to exist as actions

We can discuss these suggestions further if you like.

Thanks,
Nathan.

awset commented:
hi @salsa-nathan , so I have refactored now, 
- added access check for other actions
- moved non action function helper.
- move duplicate row builder to helper, I didn't move all of them , only the one with duplicate row, let me know if this need to be moved as well, I just don't want to do over refactoring.
- with contact point, I use the same caching strategy as packages and resources, so once it is looked up, I added it to the local var and use the pos id as key.

let me know if there are more to do. happy to do it, last night I don't spend much time to refactor it.
